poly-commit/src/hyrax/mod.rs

```diff
@@ -0,0 +1,533 @@
+    /// the commitment phase (`randomness`). Cf. sections "Square-root
+    /// commitment scheme" and appendix A.2 from the reference article.
+    ///
+    /// # Panics
+    ///
+    /// Panics if
+    /// - `rng` is None, since Hyrax requires randomness in order to
+    /// open the commitment to a polynomial.
+    /// - The point doesn't have an even number of variables.
+    /// - The labels of a commitment doesn't match that of the corresponding
+    /// polynomial.
+    /// - The number of variables of a polynomial doesn't match that of the
+    /// point.
+    fn open<'a>(
+        ck: &Self::CommitterKey,
+        labeled_polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField, P>>,
+        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        point: &'a P::Point,
+        sponge: &mut S,
+        states: impl IntoIterator<Item = &'a Self::CommitmentState>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> Result<Self::Proof, Self::Error>
+    where
+        Self::Commitment: 'a,
+        Self::CommitmentState: 'a,
+        P: 'a,
+    {
+        let n = point.len();
+
+        if n % 2 == 1 {
+            // Only polynomials with an even number of variables are
+            // supported in this implementation
+            return Err(Error::InvalidNumberOfVariables);
+        }
+
+        let dim = 1 << n / 2;
+
+        // Reversing the point is necessary because the MLE interface returns
+        // evaluations in little-endian order
+        let point_rev: Vec<G::ScalarField> = point.iter().rev().cloned().collect();
+
+        let point_lower = &point_rev[n / 2..];
+        let point_upper = &point_rev[..n / 2];
+
+        // Deriving the tensors which result in the evaluation of the polynomial
+        // when they are multiplied by the coefficient matrix.
+        let l = tensor_prime(point_lower);
+        let r = tensor_prime(point_upper);
+
+        let mut proofs = Vec::new();
+
+        let rng_inner = rng.expect("Opening polynomials requires randomness");
+
+        for (l_poly, (l_com, state)) in labeled_polynomials
+            .into_iter()
+            .zip(commitments.into_iter().zip(states.into_iter()))
+        {
+            let label = l_poly.label();
+            if label != l_com.label() {
+                return Err(Error::MismatchedLabels {
+                    commitment_label: l_com.label().to_string(),
+                    polynomial_label: label.to_string(),
+                });
+            }
+
+            let poly = l_poly.polynomial();
+            let com = l_com.commitment();
+
+            if poly.num_vars() != n {
+                return Err(Error::MismatchedNumVars {
+                    poly_nv: poly.num_vars(),
+                    point_nv: n,
+                });
+            }
+
+            // Absorbing public parameters
+            sponge.absorb(&to_bytes!(ck).map_err(|_| Error::TranscriptError)?);
```

Comment:
This will be quite slow in practice. A faster way would be compute  the hash of the generators, store this in the commitment key and absorb that hash here.

Reply:
I was not sure which part was the slow part, serializing or absorbing. I assumed the latter. Please check https://github.com/arkworks-rs/poly-commit/pull/130/commits/ac4a14c5e7a3486d845229e0ef529ed64457648e and let me know if it was what you meant